File: snippets/python/dependency-injection.py
"""
USE: When implementing dependency injection for testable, loosely-coupled code
REQUIRES: Python 3.10+
PATTERN: Configuration
"""

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from enum import Enum, auto
from typing import Any, Callable, Generic, TypeVar

T = TypeVar("T")


class Lifecycle(Enum):
    """Dependency lifecycle options."""
    TRANSIENT = auto()  # New instance every time
    SINGLETON = auto()  # Single shared instance
    SCOPED = auto()      # One instance per scope (e.g., request)


@dataclass
class Registration(Generic[T]):
    """Dependency registration metadata."""
    factory: Callable[["Container"], T]
    lifecycle: Lifecycle
    instance: T | None = None


class ContainerError(Exception):
    """Dependency injection error."""
    pass


class Container:
    """
    Simple dependency injection container.

    Supports constructor injection with lifecycle management.
    """

    def __init__(self) -> None:
        self._registrations: dict[type, Registration[Any]] = {}
        self._scoped_instances: dict[type, Any] = {}
        self._in_scope: bool = False

    def register(
        self,
        interface: type[T],
        factory: Callable[["Container"], T],
        lifecycle: Lifecycle = Lifecycle.TRANSIENT,
    ) -> "Container":
        """
        Register a dependency.

        Args:
            interface: The type/interface to register
            factory: Factory function that creates the instance
            lifecycle: Instance lifecycle (transient, singleton, scoped)

        Returns:
            Self for method chaining
        """
        self._registrations[interface] = Registration(
            factory=factory,
            lifecycle=lifecycle,
        )
        return self

    def register_instance(self, interface: type[T], instance: T) -> "Container":
        """Register an existing instance as singleton."""
        self._registrations[interface] = Registration(
            factory=lambda _: instance,
            lifecycle=Lifecycle.SINGLETON,
            instance=instance,
        )
        return self

    def register_transient(
        self,
        interface: type[T],
        factory: Callable[["Container"], T],
    ) -> "Container":
        """Register a transient dependency (new instance each time)."""
        return self.register(interface, factory, Lifecycle.TRANSIENT)

    def register_singleton(
        self,
        interface: type[T],
        factory: Callable[["Container"], T],
    ) -> "Container":
        """Register a singleton dependency (shared instance)."""
        return self.register(interface, factory, Lifecycle.SINGLETON)

    def register_scoped(
        self,
        interface: type[T],
        factory: Callable[["Container"], T],
    ) -> "Container":
        """Register a scoped dependency (one per scope)."""
        return self.register(interface, factory, Lifecycle.SCOPED)

    def resolve(self, interface: type[T]) -> T:
        """
        Resolve a dependency.

        Args:
            interface: The type/interface to resolve

        Returns:
            Instance of the requested type

        Raises:
            ContainerError: If dependency is not registered
        """
        if interface not in self._registrations:
            raise ContainerError(f"No registration found for {interface.__name__}")

        registration = self._registrations[interface]

        # Handle singleton
        if registration.lifecycle == Lifecycle.SINGLETON:
            if registration.instance is None:
                registration.instance = registration.factory(self)
            return registration.instance

        # Handle scoped
        if registration.lifecycle == Lifecycle.SCOPED:
            if not self._in_scope:
                raise ContainerError(
                    f"Cannot resolve scoped dependency {interface.__name__} outside of scope"
                )
            if interface not in self._scoped_instances:
                self._scoped_instances[interface] = registration.factory(self)
            return self._scoped_instances[interface]

        # Transient - always create new
        return registration.factory(self)

    def create_scope(self) -> "ContainerScope":
        """Create a new scope for scoped dependencies."""
        return ContainerScope(self)

    def _begin_scope(self) -> None:
        """Begin a new scope."""
        self._in_scope = True
        self._scoped_instances.clear()

    def _end_scope(self) -> None:
        """End the current scope."""
        self._in_scope = False
        self._scoped_instances.clear()


class ContainerScope:
    """Context manager for scoped dependencies."""

    def __init__(self, container: Container):
        self._container = container

    def __enter__(self) -> Container:
        self._container._begin_scope()
        return self._container

    def __exit__(self, *args: Any) -> None:
        self._container._end_scope()


# CUSTOMIZE: Define your interfaces and implementations


class ILogger(ABC):
    """Logger interface."""

    @abstractmethod
    def info(self, message: str) -> None:
        pass

    @abstractmethod
    def error(self, message: str, error: Exception | None = None) -> None:
        pass


class IUserRepository(ABC):
    """User repository interface."""

    @abstractmethod
    def get_by_id(self, user_id: str) -> dict[str, Any] | None:
        pass

    @abstractmethod
    def save(self, user: dict[str, Any]) -> None:
        pass


class IEmailService(ABC):
    """Email service interface."""

    @abstractmethod
    def send(self, to: str, subject: str, body: str) -> bool:
        pass


# CUSTOMIZE: Implement your dependencies


class ConsoleLogger(ILogger):
    """Simple console logger implementation."""

    def info(self, message: str) -> None:
        print(f"[INFO] {message}")

    def error(self, message: str, error: Exception | None = None) -> None:
        print(f"[ERROR] {message}")
        if error:
            print(f"  Exception: {error}")


class InMemoryUserRepository(IUserRepository):
    """In-memory user repository for testing."""

    def __init__(self) -> None:
        self._users: dict[str, dict[str, Any]] = {}

    def get_by_id(self, user_id: str) -> dict[str, Any] | None:
        return self._users.get(user_id)

    def save(self, user: dict[str, Any]) -> None:
        self._users[user["id"]] = user


@dataclass
class UserService:
    """Example service using injected dependencies."""
    repository: IUserRepository
    logger: ILogger
    email_service: IEmailService

    def create_user(self, name: str, email: str) -> dict[str, Any]:
        user = {"id": f"user_{name}", "name": name, "email": email}
        self.repository.save(user)
        self.logger.info(f"Created user: {name}")
        self.email_service.send(email, "Welcome!", f"Hello {name}!")
        return user


# CUSTOMIZE: Configure your container


def configure_container() -> Container:
    """Configure dependency injection container."""
    container = Container()

    # Register dependencies
    container.register_singleton(ILogger, lambda c: ConsoleLogger())
    container.register_singleton(IUserRepository, lambda c: InMemoryUserRepository())

    # CUSTOMIZE: Replace with real implementation
    # container.register_singleton(IEmailService, lambda c: SmtpEmailService(c.resolve(ILogger)))

    # Register services that depend on other services
    container.register_transient(
        UserService,
        lambda c: UserService(
            repository=c.resolve(IUserRepository),
            logger=c.resolve(ILogger),
            email_service=c.resolve(IEmailService),
        ),
    )

    return container


# CUSTOMIZE: Example usage
# container = configure_container()
# user_service = container.resolve(UserService)
# user = user_service.create_user("Alice", "alice@example.com")
#
# # Using scopes (e.g., per-request in web apps)
# with container.create_scope() as scoped:
#     request_service = scoped.resolve(RequestScopedService)

File: snippets/java/dependency-injection.java
/**
 * USE: When implementing dependency injection patterns with Spring or Jakarta EE
 * REQUIRES: Java 17+, Spring Boot 3.x or Jakarta EE 10+
 * PATTERN: Configuration, Error Handling
 */

import java.util.Objects;

// ============================================================================
// PATTERN 1: Constructor Injection (Recommended)
// ============================================================================

// CUSTOMIZE: Your service class with constructor injection
// @Service // Spring
// @ApplicationScoped // Jakarta CDI
public class UserService {
    private final UserRepository userRepository;
    private final EmailService emailService;
    private final UserMapper userMapper;

    // Constructor injection - preferred approach
    // Spring: @Autowired is optional for single constructor
    // Jakarta CDI: @Inject required
    public UserService(
            UserRepository userRepository,
            EmailService emailService,
            UserMapper userMapper) {
        this.userRepository = Objects.requireNonNull(userRepository, "userRepository");
        this.emailService = Objects.requireNonNull(emailService, "emailService");
        this.userMapper = Objects.requireNonNull(userMapper, "userMapper");
    }

    public UserDto createUser(CreateUserRequest request) {
        var user = userMapper.toEntity(request);
        var savedUser = userRepository.save(user);
        emailService.sendWelcomeEmail(savedUser.getEmail());
        return userMapper.toDto(savedUser);
    }
}

// ============================================================================
// PATTERN 2: Configuration Classes
// ============================================================================

// CUSTOMIZE: Spring configuration class
// @Configuration
public class AppConfig {

    // @Bean
    public UserMapper userMapper() {
        return new UserMapperImpl();
    }

    // @Bean
    // @Profile("!test") // Exclude from test profile
    public EmailService emailService(EmailConfig config) {
        return new SmtpEmailService(config);
    }

    // @Bean
    // @ConditionalOnProperty(name = "feature.caching.enabled", havingValue = "true")
    public CacheManager cacheManager() {
        return new CaffeineCacheManager();
    }
}

// CUSTOMIZE: Configuration properties class
// @ConfigurationProperties(prefix = "app.email")
// @Validated
public record EmailConfig(
    // @NotBlank
    String host,
    // @Min(1) @Max(65535)
    int port,
    // @NotBlank
    String username,
    String password,
    boolean sslEnabled
) {
    public EmailConfig {
        Objects.requireNonNull(host, "host cannot be null");
        Objects.requireNonNull(username, "username cannot be null");
        if (port < 1 || port > 65535) {
            throw new IllegalArgumentException("port must be between 1 and 65535");
        }
    }
}

// ============================================================================
// PATTERN 3: Interface-Based Injection with Multiple Implementations
// ============================================================================

// Define interface
public interface NotificationService {
    void send(String recipient, String message);
}

// Implementation 1
// @Service
// @Primary // Default implementation
public class EmailNotificationService implements NotificationService {
    private final EmailService emailService;

    public EmailNotificationService(EmailService emailService) {
        this.emailService = Objects.requireNonNull(emailService, "emailService");
    }

    @Override
    public void send(String recipient, String message) {
        emailService.sendEmail(recipient, "Notification", message);
    }
}

// Implementation 2
// @Service("smsNotification")
public class SmsNotificationService implements NotificationService {
    private final SmsGateway smsGateway;

    public SmsNotificationService(SmsGateway smsGateway) {
        this.smsGateway = Objects.requireNonNull(smsGateway, "smsGateway");
    }

    @Override
    public void send(String recipient, String message) {
        smsGateway.sendSms(recipient, message);
    }
}

// Using qualifier to select implementation
// @Service
public class AlertService {
    private final NotificationService primaryNotification;
    private final NotificationService smsNotification;

    public AlertService(
            // @Primary automatically selected, or use @Qualifier
            NotificationService primaryNotification,
            // @Qualifier("smsNotification")
            NotificationService smsNotification) {
        this.primaryNotification = primaryNotification;
        this.smsNotification = smsNotification;
    }

    public void sendCriticalAlert(String recipient, String message) {
        primaryNotification.send(recipient, message);
        smsNotification.send(recipient, message);
    }
}

// ============================================================================
// PATTERN 4: Factory Pattern with DI
// ============================================================================

// Factory interface
public interface PaymentProcessorFactory {
    PaymentProcessor create(PaymentMethod method);
}

// @Component
public class PaymentProcessorFactoryImpl implements PaymentProcessorFactory {
    private final CreditCardProcessor creditCardProcessor;
    private final PayPalProcessor payPalProcessor;
    private final BankTransferProcessor bankTransferProcessor;

    public PaymentProcessorFactoryImpl(
            CreditCardProcessor creditCardProcessor,
            PayPalProcessor payPalProcessor,
            BankTransferProcessor bankTransferProcessor) {
        this.creditCardProcessor = creditCardProcessor;
        this.payPalProcessor = payPalProcessor;
        this.bankTransferProcessor = bankTransferProcessor;
    }

    @Override
    public PaymentProcessor create(PaymentMethod method) {
        return switch (method) {
            case CREDIT_CARD -> creditCardProcessor;
            case PAYPAL -> payPalProcessor;
            case BANK_TRANSFER -> bankTransferProcessor;
        };
    }
}

// ============================================================================
// PATTERN 5: Lazy Initialization
// ============================================================================

// @Service
public class ReportService {
    // Inject provider for lazy initialization
    // Spring: ObjectProvider<ExpensiveService>
    // Jakarta CDI: Instance<ExpensiveService>
    private final /* ObjectProvider<ExpensiveService> */ Object expensiveServiceProvider;

    public ReportService(/* ObjectProvider<ExpensiveService> */ Object expensiveServiceProvider) {
        this.expensiveServiceProvider = expensiveServiceProvider;
    }

    public Report generateReport(ReportType type) {
        // Service only created when needed
        // var service = expensiveServiceProvider.getIfAvailable();
        // return service != null ? service.generate(type) : Report.empty();
        return null; // Replace with actual implementation
    }
}

// ============================================================================
// PATTERN 6: Scope Management
// ============================================================================

// Request-scoped bean (new instance per HTTP request)
// @RequestScope
// @Component
public class RequestContext {
    private String correlationId;
    private String userId;

    public void initialize(String correlationId, String userId) {
        this.correlationId = correlationId;
        this.userId = userId;
    }

    public String getCorrelationId() { return correlationId; }
    public String getUserId() { return userId; }
}

// Prototype-scoped bean (new instance each time injected)
// @Scope("prototype")
// @Component
public class TransactionContext {
    private final String transactionId;

    public TransactionContext() {
        this.transactionId = java.util.UUID.randomUUID().toString();
    }

    public String getTransactionId() { return transactionId; }
}

// ============================================================================
// Stub interfaces for compilation (remove in actual usage)
// ============================================================================
interface UserRepository { Object save(Object user); }
interface EmailService {
    void sendWelcomeEmail(String email);
    void sendEmail(String to, String subject, String body);
}
interface UserMapper {
    Object toEntity(CreateUserRequest request);
    UserDto toDto(Object user);
}
record CreateUserRequest(String name, String email) {}
record UserDto(String id, String name, String email) {}
interface CacheManager {}
class CaffeineCacheManager implements CacheManager {}
class UserMapperImpl implements UserMapper {
    public Object toEntity(CreateUserRequest request) { return null; }
    public UserDto toDto(Object user) { return null; }
}
class SmtpEmailService implements EmailService {
    SmtpEmailService(EmailConfig config) {}
    public void sendWelcomeEmail(String email) {}
    public void sendEmail(String to, String subject, String body) {}
}
interface SmsGateway { void sendSms(String to, String message); }
interface PaymentProcessor {}
class CreditCardProcessor implements PaymentProcessor {}
class PayPalProcessor implements PaymentProcessor {}
class BankTransferProcessor implements PaymentProcessor {}
enum PaymentMethod { CREDIT_CARD, PAYPAL, BANK_TRANSFER }
record Report() { static Report empty() { return new Report(); } }
enum ReportType { DAILY, WEEKLY, MONTHLY }
